File: solutions/Leetcode_1448/leetcode_1448.py
from leetcode.python.tree import TreeNode


class Solution1:
    def good_nodes(self, root: TreeNode) -> int:
        count = 0
        stack = [(root, -1e4-1)]
        while stack:
            node, max_value = stack.pop()
            if node:
                if node.val >= max_value:
                    count += 1
                    max_value = node.val
                stack.extend([(node.left, max_value), (node.right, max_value)])
        return count

File: solutions/Leetcode_1448/leetcode_1448.cpp
#include <stack>
#include <utility>

class Solution1
{
public:
    int goodNodes(TreeNode *root)
    {
        int count = 0;
        std::stack<std::pair<TreeNode *, int>> s;
        s.push({root, -10000});
        TreeNode *node;
        int maxValue;
        while (!s.empty())
        {
            std::pair<TreeNode *, int> nodeMaxValue = s.top();
            node = nodeMaxValue.first;
            maxValue = nodeMaxValue.second;
            s.pop();
            if (node != nullptr)
            {
                if (node->val >= maxValue)
                {
                    maxValue = node->val;
                    count++;
                }
                s.push({node->left, maxValue});
                s.push({node->right, maxValue});
            }
        }
        return count;
    }
};

class Solution1Recursion
{
private:
    int count = 0;

    void dfs(TreeNode *node, int maxValue)
    {
        if (node != nullptr)
        {
            if (node->val >= maxValue)
            {
                count++;
                maxValue = node->val;
            }
            dfs(node->left, maxValue);
            dfs(node->right, maxValue);
        }
    }

public:
    int goodNodes(TreeNode *root)
    {
        dfs(root, -10000);
        return count;
    }
};